onnx/reference/ops/op_cast.py

```diff
@@ -74,6 +77,25 @@ def cast_to(x, to, saturate):  # noqa: PLR0911
             y[i] = el
         return y.reshape(x.shape)
 
+    i4 = [
+        (uint4, "uint4", TensorProto.UINT4, False),
+        (int4, "int4", TensorProto.INT4, True),
+    ]
+    for np_type, np_desc, tensor_type, signed in i4:
+        if x.dtype == np_type and x.dtype.descr[0][0] == np_desc:
+            if to == tensor_type:
```

Comment:
I don't where tensor_type is defined.

Reply:
Just in the row above 
`    i4 = [
        (uint4, "uint4", TensorProto.UINT4, False),
        (int4, "int4", TensorProto.INT4, True),
    ]`
 Then `np_type, np_desc, tensor_type, signed` are unpacked from the tuple. 
It is similar to the handling of fp8 types in lines 45-60 `for (dt, st, proto_type), cvt in f8.items(): `

I admit it's not the most elegant solution.

Perhaps we can add a map in `custom_element_types.py` between the numpy type, the TenorProto type and whatever other attributes are needed.